File: update/update.go
package update

import (
    "fmt"
    "math"
    "math/rand"
    "github.com/jhpeng/ctQMC/update/cluster"
    . "github.com/jhpeng/ctQMC/dtype"
)

var cmf []float64
var count []uint64
func pointSamplingWeighted(t float64, w []float64) int {
    dis := rand.Float64()*t

    if cmf==nil {
        cmf    = make([]float64, len(w))
        count  = make([]uint64,  len(w))
        cmf[0] = w[0]
        count[0] = 0
        for i:=1;i<len(w);i++ {
            cmf[i] = cmf[i-1]+w[i]
            count[i] = 0
        }
    }

    i:=-1
    j:=len(cmf)
    d:=j-i
    for d>1 {
        if dis > cmf[i+d/2] {
            i = i+d/2
        } else {
            j = i+d/2
        }
        d = j-i
    }

    count[j]++

    return j
}

var MaxKeyFromSample uint64 = 1000000000000000 // 10E15
func squenceSamplingUniform(lambda float64, start uint64) []uint64{
    var sequence []uint64
    var k float64 = 0

    dis := rand.Float64()
    k -= math.Log(dis)/lambda
    for k<1.0 {
        sequence = append(sequence,uint64(k*float64(MaxKeyFromSample))+start)
        dis = rand.Float64()
        k -= math.Log(dis)/lambda
    }

    return sequence
}

func NewWorldLine(mnspin int, nsite int, beta float64) WorldLine {
    var w WorldLine

    length := 1024
    w.SequenceA = make([]Vertex,length)
    w.SequenceB = make([]Vertex,length)
    w.Mnspin    = mnspin
    w.Cluster   = make([]int,mnspin*length)
    w.Weight    = make([]int,mnspin*length)

    w.Nvertices = 0
    w.Flag  = true
    w.State = make([]int,nsite)
    w.Last  = make([]int,nsite)
    w.First = make([]int,nsite)
    w.Nsite = nsite
    w.Beta = beta

    return w
}

func Remove(w WorldLine) WorldLine{
    var v   Vertex

    sequence1 := w.SequenceB
    sequence2 := w.SequenceA
    if w.Flag {
        sequence1 = w.SequenceA
        sequence2 = w.SequenceB
    }

    k:=0
    for i:=0;i<w.Nvertices;i++ {
        v = sequence1[i]
        check_delete := true

        for j:=0;j<v.HNspin;j++ {
            if v.State[j]!=v.State[j+v.HNspin]{
                check_delete = false
            }
        }

        if !check_delete {
            sequence2[k] = sequence1[i]
            k++
        }
    }
    w.Nvertices = k
    w.Flag = !(w.Flag)

    return w
}

func Insert(w WorldLine, m Model) WorldLine{
    lambda := (m.Tweight)*(w.Beta)
    insert_seq := squenceSamplingUniform(lambda,0)

    length := len(insert_seq)+w.Nvertices+1024
    if length>len(w.SequenceA) {
        temp1 := make([]Vertex,length)
        copy(temp1,w.SequenceA)
        temp2 := make([]Vertex,length)
        copy(temp2,w.SequenceB)
        w.SequenceA = temp1
        w.SequenceB = temp2

        w.Cluster = make([]int,length*w.Mnspin)
        w.Weight  = make([]int,length*w.Mnspin)
    }

    sequence1 := w.SequenceB
    sequence2 := w.SequenceA
    if w.Flag {
        sequence1 = w.SequenceA
        sequence2 = w.SequenceB
    }

    pstate := make([]int,m.Nsite)
    copy(pstate,w.State)

    k:=0
    n:=0
    var v Vertex
    for i:=0;i<len(insert_seq);i++ {
        key1 := (sequence1[k]).Key
        key2 := insert_seq[i]

        for (key1<key2) && (k<w.Nvertices) {
            v = sequence1[k]
            indices := m.Bond2index[v.Bond]
            for i_site:=0;i_site<v.HNspin;i_site++ {
                index := indices[i_site]
                pstate[index] = v.State[v.HNspin+i_site]
            }

            sequence2[n] = v
            n++
            k++
            key1 = (sequence1[k]).Key
        }

        if key1!=key2 {
            bond    := pointSamplingWeighted(m.Tweight,m.Bond2weight)
            t       := m.Bond2type[bond]
            indices := m.Bond2index[bond]
            hNspin  := m.Bond2hNspin[bond]
            rule    := m.InsertRule[t]
            lstate  := make([]int,2*hNspin)
            for i_site:=0;i_site<hNspin;i_site++ {
                lstate[i_site]        = pstate[indices[i_site]]
                lstate[i_site+hNspin] = pstate[indices[i_site]]
            }

            if rule(lstate) {
                var iv Vertex
                iv.Key    = key2
                iv.Bond   = bond
                iv.HNspin = hNspin
                iv.State  = lstate

                sequence2[n] = iv
                n++
            }
        }
    }

    for k<w.Nvertices {
        sequence2[n] = sequence1[k]
        n++
        k++
    }

    w.Nvertices = n
    w.Flag = !(w.Flag)

    return w
}

func InnerLink(w WorldLine, m Model) {

    var (
        v      Vertex
        bond   int
        hNspin int
        t      int
        idn    int
        idv    int
    )

    sequence := w.SequenceB
    if w.Flag {
        sequence = w.SequenceA
    }

    for i:=0;i<w.Nvertices;i++ {
        v      = sequence[i]
        bond   = v.Bond
        hNspin = v.HNspin
        t      = m.Bond2type[bond]
        rule  := m.LinkRule[t]

        for j:=0;j<2*hNspin;j++ {
            idn = i*w.Mnspin+j
            idv = i*w.Mnspin+rule[j]
            w.Cluster[idn] = idv
            w.Weight[idn]  = rule[j+2*hNspin]
        }
    }
}

func OuterLink(w WorldLine, m Model) {

    var (
        v         Vertex
        bond      int
        hNspin    int
        indices []int
        index     int
        idp       int
        idn       int
    )

    last  := w.Last
    first := w.First

    for i:=0;i<w.Nsite;i++ {
        last[i]  = -1
        first[i] = -1
    }

    sequence := w.SequenceB
    if w.Flag {
        sequence = w.SequenceA
    }

    for i:=0;i<w.Nvertices;i++ {
        v       = sequence[i]
        bond    = v.Bond
        hNspin  = v.HNspin
        indices = m.Bond2index[bond]

        for j:=0;j<hNspin;j++ {
            index = indices[j]
            idp = i*w.Mnspin+j
            idn = i*w.Mnspin+j+hNspin
            if first[index]==-1 {
                first[index] = idp
                last[index]  = idn
            } else {
                cluster.Union(w.Cluster,w.Weight,last[index],idp)
                last[index] = idn
            }
        }
    }

    for i:=0;i<w.Nsite;i++ {
        if first[i]!=-1 {
            cluster.Union(w.Cluster,w.Weight,first[i],last[i])
        }
    }
}

func FlipCluster(w WorldLine) {

    var(
        v       Vertex
        hNspin  int
        state []int
        idv     int
        idr     int
        id      int
        p       int
        i       int
        j       int
    )

    newState := make(map[int]int)
    first    := w.First

    sequence := w.SequenceB
    if w.Flag {
        sequence = w.SequenceA
    }

    for i=0;i<w.Nvertices;i++ {
        v       = sequence[i]
        hNspin  = v.HNspin
        state   = v.State

        for j=0;j<2*hNspin;j++ {
            idv = i*w.Mnspin+j
            idr = cluster.Root(w.Cluster,idv)
            if newState[idr]==0 {
                if rand.Float64()<0.5 {
                    newState[idr]= 1
                } else {
                    newState[idr]=-1
                }
            }

            state[j] = state[j]*newState[idr]
        }
    }

    for i=0;i<w.Nsite;i++ {
        id  = first[i]
        if id!=-1 {
            p = id/w.Mnspin
            j = id%w.Mnspin

            v = sequence[p]
            w.State[i] = v.State[j]
        } else if rand.Float64()<0.5 {
            w.State[i] =  1
        } else {
            w.State[i] = -1
        }
    }
}

func CheckPeriodic(w WorldLine, m Model) bool {
    sequence := w.SequenceB
    if w.Flag {
        sequence = w.SequenceA
    }

    pstate := make([]int,w.Nsite)
    for i:=0;i<w.Nsite;i++ {
        pstate[i] = w.State[i]
    }

    for i:=0;i<w.Nvertices;i++ {
        v       := sequence[i]
        bond    := v.Bond
        hNspin  := v.HNspin
        indices := m.Bond2index[bond]

        for j:=0;j<hNspin;j++ {
            i_site := indices[j]
            if pstate[i_site]!=v.State[j]{
                fmt.Printf("something wrong!\n")
            }
            pstate[i_site] = v.State[hNspin+j]
        }
    }

    check := true
    for i:=0;i<w.Nsite;i++ {
        if pstate[i]!=w.State[i] {
            check = false
        }
    }

    return check
}

File: C/update.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <gsl/gsl_rng.h>

#include "dtype.h"
#include "union_find.h"

//#define weighted_sampling

#ifdef weighted_sampling
static int weighted_sampling(double* cmf, int length, gsl_rng* rng) {
    double dis = gsl_rng_uniform_pos(rng)*cmf[length-1];

    int i = -1;
    int j = length;
    int d = j-1;

    while(d>1) {
        if(dis>cmf[i+d/2]){
            i = i+d/2;
        } else {
            j = i+d/2;
        }
        d = j-i;
    }

    return j;
}
#endif

static double* insert_seq;
static int* insert_bond;
static int insert_len=0;
static int insert_cap=0;
static double d_ave=-1;
static double d_max=-1;
static void uniform_sequence_sampling(model* m, double lam, double start, gsl_rng* rng) {
    if(d_max<0){
        d_max = 0;
        d_ave = 0;
        for(int i=0;i<(m->nbond);i++) {
            d_ave += m->bond2weight[i];
            if((m->bond2weight[i])>d_max)
                d_max = m->bond2weight[i];
        }
        d_ave = d_ave/(m->nbond);
    }

    if(insert_cap==0) {
        insert_cap = (int)(lam+sqrt(lam)*10+1024);
        insert_seq  = (double*)malloc(sizeof(double)*insert_cap);
        insert_bond = (int*)malloc(sizeof(int)*insert_cap);
    }

    lam = lam*d_max/d_ave;

    double k=0;
    int n=0;
    int bond;

    double dis = gsl_rng_uniform_pos(rng);
    k -= log(dis)/lam;
    while((k<1.0) && (n<insert_cap)) {
        bond = gsl_rng_uniform_pos(rng)*(m->nbond);
        if(gsl_rng_uniform_pos(rng)*d_max<(m->bond2weight[bond])){
            insert_seq[n]  = k+start;
            insert_bond[n] = bond;
            n++;
        }

        dis = gsl_rng_uniform_pos(rng);
        k -= log(dis)/lam;
    }
    while(k<1.0) {
        dis = gsl_rng_uniform_pos(rng);
        k -= log(dis)/lam;
        n++;
    }
    insert_len = n;
    if(n>insert_cap) {
        double* seq = (double*)malloc(sizeof(double)*n*2);
        int* bond   = (int*)malloc(sizeof(int)*n*2);
        for(int i=0;i<insert_cap;i++){
            seq[i]  = insert_seq[i];
            bond[i] = insert_bond[i];
        }
        free(insert_seq);
        insert_seq  = seq;
        insert_bond = bond;
        insert_len = insert_cap;
        insert_cap = n*2;
    }
}

void remove_vertices(world_line* w) {
    vertex* v;

    vertex* sequence1 = w->sequenceB;
    vertex* sequence2 = w->sequenceA;
    if(w->flag) {
        sequence1 = w->sequenceA;
        sequence2 = w->sequenceB;
    }

    int check_delete;
    int i,j,k;
    k=0;
    for(i=0;i<w->nvertices;i++) {
        v = &(sequence1[i]);
        check_delete = 1;

        for(j=0;j<(v->hNspin);j++) {
            if(v->state[j]!=v->state[j+v->hNspin])
                check_delete = 0;
        }

        if(!check_delete) {
            copy_vertex(&(sequence2[k]),&(sequence1[i]));
            k++;
        }
    }
    w->nvertices = k;
    w->flag = !(w->flag);
}

void insert_vertices(world_line* w, model* m, gsl_rng* rng) {
    double lam = (m->sweight)*(w->beta);

    uniform_sequence_sampling(m,lam,0,rng);

    int length = insert_cap+w->nvertices;
    realloc_world_line(w,length);

    vertex* sequence1 = w->sequenceB;
    vertex* sequence2 = w->sequenceA;
    if(w->flag) {
        sequence1 = w->sequenceA;
        sequence2 = w->sequenceB;
    }

    int* pstate = w->pstate;
    int  nsite = w->nsite;

    for(int i=0;i<nsite;i++) pstate[i] = w->istate[i];

    vertex* v;
    int n,i,k,i_site,index;
    double tau1,tau2;

    int mhnspin = m->mhnspin;
    int lstate[mhnspin];

    k=0;
    n=0;

    tau1 = 0;
    if(w->nvertices!=0) tau1 = (sequence1[0]).tau;

    for(i=0;i<insert_len;i++) {
        tau2 = insert_seq[i];

        while((tau1<tau2) && (k<(w->nvertices))) {
            v = &(sequence1[k]);
            for(i_site=0;i_site<(v->hNspin);i_site++) {
                index = m->bond2index[v->bond*mhnspin+i_site];
                pstate[index] = v->state[v->hNspin+i_site];
            }

            copy_vertex(&(sequence2[n]),v);
            n++;
            k++;

            if(k<(w->nvertices)){
                tau1 = (sequence1[k]).tau;
            }
        }

        if(tau1!=tau2) {
            int bond     = insert_bond[i];
            int t        = m->bond2type[bond];
            int hNspin   = m->bond2hNspin[bond];
            insert_rule rule = m->insert[t];

            for(i_site=0;i_site<hNspin;i_site++) { 
                index = m->bond2index[bond*mhnspin+i_site];
                lstate[i_site] = pstate[index];
            }

            if(rule(lstate)) {
                (sequence2[n]).tau    = tau2;
                (sequence2[n]).bond   = bond;
                (sequence2[n]).hNspin = hNspin;

                for(i_site=0;i_site<hNspin;i_site++) {
                    (sequence2[n]).state[i_site]        = lstate[i_site];
                    (sequence2[n]).state[i_site+hNspin] = lstate[i_site];
                }

                n++;
            }
        }
    }

    while(k<(w->nvertices)) {
        copy_vertex(&(sequence2[n]),&(sequence1[k]));
        n++;
        k++;
    }

    w->nvertices = n;
    w->flag = !(w->flag);
}

void clustering(world_line* w, model* m) {
    vertex* v;
    int bond,hNspin,t,idn,idp;
    int i,j,index;
    int mnspin = w->mnspin;
    int nsite = w->nsite;
    int* rule;
    int* indices;

    int* first = w->first;
    int* last  = w->last;

    for(i=0;i<nsite;i++) {
        last[i]  = -1;
        first[i] = -1;
    }

    vertex* sequence = w->sequenceB;
    if(w->flag) 
        sequence = w->sequenceA;

    for(i=0;i<(w->nvertices);i++) {
        v       = &(sequence[i]);
        bond    = v->bond;
        hNspin  = v->hNspin;
        t       = m->bond2type[bond];
        rule    = &(m->link[4*(m->mhnspin)*t]);
        indices = &(m->bond2index[bond*(m->mhnspin)]);

        for(j=0;j<2*hNspin;j++) {
            idn = i*mnspin+j;
            idp = i*mnspin+rule[j];
            w->cluster[idn] = idp;
            w->weight[idn]  = rule[2*hNspin+j];
        }

        for(j=0;j<hNspin;j++) {
            index = indices[j];
            idp = i*mnspin+j;
            idn = i*mnspin+j+hNspin;
            if(first[index]==-1) {
                first[index] = idp;
                last[index]  = idn;
            } else {
                merge(w->cluster,w->weight,last[index],idp);
                last[index] = idn;
            }
        }
    }

    for(i=0;i<nsite;i++) {
        if(first[i]!=-1) {
            merge(w->cluster,w->weight,first[i],last[i]);
        }
    }
}

void flip_cluster(world_line* w, gsl_rng* rng) {
    int* state;
    int hNspin,idv,idr,id,p,i,j;

    int mnspin = w->mnspin;
    int nsite  = w->nsite;

    vertex* sequence = w->sequenceB;
    if(w->flag) 
        sequence = w->sequenceA;

    for(i=0;i<(w->nvertices);i++) {
        hNspin = (sequence[i]).hNspin;
        state  = (sequence[i]).state;

        for(j=0;j<2*hNspin;j++) {
            idv = i*mnspin+j;
            idr = root(w->cluster,idv);
            if(w->weight[idr]>0) {
                if(gsl_rng_uniform_pos(rng)<0.5) {
                    w->weight[idr] =  0;
                } else {
                    w->weight[idr] = -1;
                }
            }

            state[j] = state[j]*((w->weight[idr])*2+1);
        }
    }

    for(i=0;i<nsite;i++) {
        id = w->first[i];
        if(id!=-1) {
            p = id/mnspin;
            j  =id%mnspin;
            w->istate[i] = (sequence[p]).state[j];
        } else if(gsl_rng_uniform_pos(rng)<0.5) {
            w->istate[i] =  1;
        } else {
            w->istate[i] = -1;
        }
    }
}

int check_periodic(world_line* w, model* m) {
    vertex* sequence = w->sequenceB;
    if(w->flag)
        sequence = w->sequenceA;

    for(int i=0;i<(w->nsite);i++)
        w->pstate[i] = w->istate[i];

    vertex* v;
    for(int i=0;i<(w->nvertices);i++) {
        v = &(sequence[i]);
        int bond   = v->bond;
        int hNspin = v->hNspin;

        for(int j=0;j<hNspin;j++) {
            int i_site = m->bond2index[bond*(m->mhnspin)+j];
            if(w->pstate[i_site]!=(v->state[j])){
                printf("something wrong!\n");
            }
            w->pstate[i_site] = v->state[j+hNspin];
        }
    }

    int check=1;
    for(int i=0;i<(w->nsite);i++){
        if((w->pstate[i])!=(w->istate[i])) {
            check=0;
        }
    }

    return check;
}
